### PR TITLE
[bazel] GPU-support: add @local_config_cuda and @cuda

### PR DESCRIPTION
**Pull request by vors**
## Context

We take the first step at tackling the GPU-bazel support by adding bazel external workspaces `@local_config_cuda` and `@cuda`, where the first one has some hardcoded values and lists of files, and the second one provides a nicer, high-level wrapper that maps into the already expected by pytorch bazel targets that are guarded with `if_cuda` macro.

The prefix `local_config_` signifies the fact that we are breaking the bazel hermeticity philosophy by explicitly relaying on the CUDA installation that is present on the machine.

## Testing

Notice an important scenario that is unlocked by this change: compilation of cpp code that depends on cuda libraries (i.e. cuda.h and so on).

Before:
```
sergei.vorobev@cs-sv7xn77uoy-gpu-1628706590:~/src/pytorch4$ bazelisk build --define=cuda=true //:c10
ERROR: /home/sergei.vorobev/src/pytorch4/tools/config/BUILD:12:1: no such package 'tools/toolchain': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /home/sergei.vorobev/src/pytorch4/tools/toolchain and referenced by '//tools/config:cuda_enabled_and_capable'
ERROR: While resolving configuration keys for //:c10: Analysis failed
ERROR: Analysis of target '//:c10' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.259s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (2 packages loaded, 2 targets configured)
```

After: 
```
sergei.vorobev@cs-sv7xn77uoy-gpu-1628706590:~/src/pytorch4$ bazelisk build --define=cuda=true //:c10
INFO: Analyzed target //:c10 (6 packages loaded, 246 targets configured).
INFO: Found 1 target...
Target //:c10 up-to-date:
  bazel-bin/libc10.lo
  bazel-bin/libc10.so
INFO: Elapsed time: 0.617s, Critical Path: 0.04s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
```

The `//:c10` target is a good testing one for this, because it has such cases where the [glob is different](https://github.com/pytorch/pytorch/blob/075024b9a34904ec3ecdab3704c3bcaa329bdfea/BUILD.bazel#L76-L81), based on do we compile for CUDA or not.

## What is out of scope of this PR

This PR is a first in a series of providing the comprehensive GPU bazel build support. Namely, we don't tackle the [cu_library](https://github.com/pytorch/pytorch/blob/11a40ad915d4d3d8551588e303204810887fcf8d/tools/rules/cu.bzl#L2) implementation here. This would be a separate large chunk of work.

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Review by malfet — CHANGES_REQUESTED** [review]
PR itself looks good to me, but at least right now it seem that build fails in CI, isn't it?
```
 /var/lib/jenkins/.cache/bazel/_bazel_jenkins/fdf6d09bf4b4f04a71e2a7dfceb40620/external/local_config_cuda/cuda/BUILD:442:1: declared output 'external/local_config_cuda/cuda/include/cudnn.h' is a dangling symbolic link
```

File: third_party/tensorflow_cuda_bazel_build/cuda/BUILD [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+    tags = ["no-cache"],
+)
+
+genrule(
+    name = "cudnn-include",
+    outs = [
+        "include/cudnn.h",
+    ],
+    cmd = """
+        ln -s /usr/include/cudnn.h $(@D)/cudnn.h""",
```
Comment (vors) [inline third_party/tensorflow_cuda_bazel_build/cuda/BUILD:448]: we need to find where the `cudnn.h` file located on the CI machines.

**Comment by vors** [comment]
Yeap, do you know what `cudnn` version is installed on the CI machines? I tried few well-known alternatives for the `cudnn.h` location. The most generic one being the `/usr/include/cudnn.h`, see here https://github.com/pytorch/pytorch/pull/63604/files#r693126778

**Comment by malfet** [comment]
As bazel builds are currently CPU-only, CI runs are executed in `pytorch-linux-xenial-py3.6-gcc7` container.
Please change it to  `pytorch-linux-bionic-cuda10.2-cudnn7-py3.9-gcc7` in 
https://github.com/pytorch/pytorch/blob/34ed16ffef80cf11c86a1f48d1b4930b71e19866/.github/scripts/generate_ci_workflows.py#L483
And regenerate workflow by running `make generate-gha-workflows`

**Review by malfet — APPROVED** [review]

**Comment by thuyen** [comment]
For the next step: would we use self-contained package like `https://github.com/tensorflow/runtime/tree/master/third_party/rules_cuda` or writing our own rules?

**Comment by vors** [comment]
@thuyen the `@ local_config_cuda` is based on the TF rules_cuda. The hard part about writing them is that we need to generate the list of files for the local system. I don't think that we need to pull all the heavy machinery that TF has build to generate them on the fly and can just write something simpler (TF has it's own needs that are not common with pytorch).
If you find that it's easily portable, please, contribute it and feel free to replace the `@local_config_cuda`.

**Pull request closed without merging**